File: Java/src/com/venus/Tree/BSTree.java
package com.venus.Tree;

import com.venus.Tree.printer.BinaryTreeInfo;
import java.util.Comparator;

@SuppressWarnings("ALL")
public class BSTree<E> extends BinaryTree<E>{

    protected Comparator comparator;

    public BSTree() {
        this(null);
    }

    public BSTree(Comparator comparator) {
        this.comparator = comparator;
    }



    public void add(E element) {
        checkNullElement(element);

        Node<E> newNode = createNode(element, null);
        if (root == null){
            root = newNode;
            afterAdd(newNode);
            size++;
            return;
        }

        Node<E> node = root;
        Node<E> parent = root;
        int cmp = 0;
        while (node != null) {
            parent = node;
            cmp = compare(node.element, element);

            if (cmp < 0) {
                node = node.right;
            }else if (cmp > 0) {
                node = node.left;
            }else {
                node.element = element;
                return;
            }
        }

        if (cmp < 0) {
            parent.right = newNode;
        }else {
            parent.left = newNode;
        }
        newNode.parent = parent;
        afterAdd(newNode);
        size++;
    }

    public void afterAdd(Node<E> node) {

    }

    public void remove(E element) {
        remove(node(element));
    }

    private void remove(Node<E> node) {
        if (node == null) return;

        if (node.hasTwoChildren()) {
            Node<E> s = successor(node);
            node.element = s.element;
            node = s;
        }

        Node<E> replace = node.left != null ? node.left : node.right;
        if (replace != null) {
            replace.parent = node.parent;
            if (node.parent == null) {
                root = replace;
            } else if (node == node.parent.left) {
                node.parent.left = replace;
            } else {
                node.parent.right = replace;
            }
            afterRemove(replace);
        }else if (node.parent == null){
            root = null;
            afterRemove(node);
        } else {
            if (node.parent.left == node) {
                node.parent.left = null;
            } else {
                node.parent.right = null;
            }
            afterRemove(node);
        }
        size--;
    }

    public void afterRemove(Node<E> node){

    }

    private Node<E> node(E element){
        Node<E> node = root;
        while (node != null) {
            int cmp = compare(element, node.element);
            if (cmp > 0) {
                node = node.right;
            }else if (cmp < 0) {
                node = node.left;
            }else {
                return node;
            }
        }
        return null;
    }

    public boolean contains(E element){
        return node(element) != null;
    }

    protected int compare(E e1, E e2) {
        if (comparator != null){
            return comparator.compare(e1, e2);
        }
        return ((Comparable<E>)e1).compareTo(e2);
    }
}

File: Python/Tree/bstree.py
## written by xiongbiao
## date 2020-5-19

from Tree.tree import BinaryTree, Node

class BinarySearchTree(BinaryTree):

    def __init__(self):
        super().__init__()


    def add_value(self, value):

        self._check_value(value)
        new_node = self.create_node(value, None)
        if self._root is None:
            self._root = new_node
            self.after_add(new_node)
            return

        node = self._root
        parent = self._root
        while node:
            parent = node
            if node.value > value:
                node = node.left
            elif node.value < value:
                node = node.right
            else:
                node.value = value
                return
        if parent.value > value:
            parent.left = new_node
        else:
            parent.right = new_node
        new_node.parent = parent

        self.after_add(new_node)
        self._size += 1

    def after_add(self, node):
        pass


    def remove_value(self, value):
        self._remove_node(self._get_node(value))

    def _remove_node(self, node):
        if node is None:
            return None

        if node.has_two_children():
            replace = self.successor(node)
            node.value = replace.value
            node = replace

        if node.left:
            child = node.left
        else:
            child = node.right

        if child:
            child.parent = node.parent
            if node.parent.left == node:
                node.parent.left = child
            elif node.parent.right == node:
                node.parent.right = child
            else:
                self._root = child
            self.after_remove(child)
        elif node.parent is None:
            self._root = None
            self.after_remove(node)
        else:
            if node.is_left_child():
                node.parent.left = None
            elif node.is_right_child():
                node.parent.right = None
            self.after_remove(node)

        self._size -= 1

    def after_remove(self, node):
        pass

    def create_node(self, value, parent):
        return Node(value, parent)

    def _get_node(self, value):

        node = self._root
        while node:
            if node.value > value:
                node = node.left
            elif node.value < value:
                node = node.right
            else:
                return node
        return None

    def _check_value(self, value):
        if value is None:
            raise ValueError('node value cannot be None')

# nodes = [7, 4, 9, 2, 5, 8, 11, 3]
# tree = BinarySearchTree()
# for n in nodes:
#     tree.add_value(n)
#
# # print(tree.inorder_traversal('recursive'))
# # print(tree.preorder_traversal('recursive'))
# # print(tree.postorder_traversal('recursive'))
# # tree.reverse()
# tree.remove_value(7)
# tree.remove_value(9)
# tree.print()